docs/syntax.rst

```diff
@@ -29,102 +25,71 @@ The return value will be automatically casted into the hinted type. e.g.,
     res = add_xy(2.3, 1.1)
     print(res)  # 3, since return type is ti.i32
 
-.. note::
-    For differentiable programming kernels should better have either serial statements or a single parallel for-loop. If you don't use differentiable programming, feel free to ignore this tip.
 
 .. note::
 
-    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or `ti.Vector`` is not supported. Python-style tuple return is not supported. e.g.:
+    For now, we only support one scalar as return value. Returning ``ti.Matrix`` or ``ti.Vector`` is not supported. Python-style tuple return is not supported either. For example,
 
     .. code-block:: python
 
         @ti.kernel
         def bad_kernel() -> ti.Matrix:
-            return ti.Matrix([[1, 0], [0, 1]])  # ERROR!
+            return ti.Matrix([[1, 0], [0, 1]])  # Error
 
         @ti.kernel
         def bad_kernel() -> (ti.i32, ti.f32):
             x = 1
             y = 0.5
-            return x, y  #  ERROR!
-
-.. note::
-  For correct gradient behaviors in differentiable programming, please refrain from using kernel return values. Instead, store the result into a global variable (e.g. ``loss[None]``).
-
-(TODO: move the following to advanced topics)
-
-* For differentiable programming kernels should better have either serial statements or a single parallel for-loop. If you don't use differentiable programming, feel free to ignore this tip.
-
-.. code-block:: python
-
-        @ti.kernel
-        def a_hard_kernel_to_auto_differentiate():
-          sum = 0
-          for i in x:
-            sum += x[i]
-          for i in y:
-            y[i] = sum
+            return x, y  # Error
 
-        # instead, split it into multiple kernels to be nice to the Taichi autodiff compiler:
 
-        @ti.kernel
-        def reduce():
-          for i in x:
-            sum[None] += x[i]
+We also support **template arguments** (see :ref:`template_metaprogramming`) and **external array arguments** (see :ref:`external`) in Taichi kernels.
```

Comment:
Let's move this line to L15 (below the `kernel can have arguments, arguments type-hinted...`).

Reply:
I would actually stick to this organization because moving this ahead might cause a digression for the reader, who are simply learning basic kernel usages. WDYT?